From: Jan Luebbe <jlu@pengutronix.de>
Date: Sat, 20 Jul 2013 14:27:34 +0200
Subject: [PATCH] Release distrokit/beaglebone/20130720

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 183eff3..0bef6ba 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 3
 PATCHLEVEL = 8
 SUBLEVEL = 13
-EXTRAVERSION =
+EXTRAVERSION =-20130720
 NAME = Displaced Humerus Anterior
 
 # *DOCUMENTATION*
