From: Michael Grzeschik <m.grzeschik@pengutronix.de>
Date: Wed, 20 Nov 2013 12:14:45 +0100
Subject: [PATCH] Release 3.12/customers/ptx/bbb/20131120-1

Signed-off-by: Michael Grzeschik <m.grzeschik@pengutronix.de>
---
 Makefile | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index 67077ad..3d9beb5 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 3
 PATCHLEVEL = 12
 SUBLEVEL = 0
-EXTRAVERSION =
+EXTRAVERSION =-20131120-1
 NAME = One Giant Leap for Frogkind
 
 # *DOCUMENTATION*
