From 3c44683c4f806eedf920704b768f178221e652e1 Mon Sep 17 00:00:00 2001
From: Jan Luebbe <jlu@pengutronix.de>
Date: Sun, 14 Jul 2013 13:21:00 +0200
Subject: [PATCH 04/17] beaglebone: fix booting

Signed-off-by: Jan Luebbe <jlu@pengutronix.de>
---
 arch/arm/boards/beaglebone/lowlevel.c |   17 ++++++++++++++---
 1 file changed, 14 insertions(+), 3 deletions(-)

diff --git a/arch/arm/boards/beaglebone/lowlevel.c b/arch/arm/boards/beaglebone/lowlevel.c
index 28959ff..2a8e677 100644
--- a/arch/arm/boards/beaglebone/lowlevel.c
+++ b/arch/arm/boards/beaglebone/lowlevel.c
@@ -1,6 +1,7 @@
 #include <init.h>
 #include <sizes.h>
 #include <io.h>
+#include <asm/armlinux.h>
 #include <asm/barebox-arm-head.h>
 #include <asm/barebox-arm.h>
 #include <mach/am33xx-silicon.h>
@@ -201,6 +202,9 @@ void beaglebone_sram_init(void)
 
 	beaglebone_config_ddr();
 
+	/* Enable pin mux */
+	am33xx_enable_uart0_pin_mux();
+
 	/* UART softreset */
 	uart_base = AM33XX_UART0_BASE;
 
@@ -242,17 +246,24 @@ static int beaglebone_board_init(void)
 	if (!in_sdram)
 		beaglebone_sram_init();
 
-	/* Enable pin mux */
-	am33xx_enable_uart0_pin_mux();
+	return 0;
+}
+
+static int beaglebone_sram_mem_init(void)
+{
+	if (!running_in_sdram())
+		arm_add_mem_device("sram", 0x402F0400, 0x4030CE00-0x402F0400);
 
 	return 0;
 }
+mem_initcall(beaglebone_sram_mem_init);
 
-void __naked barebox_arm_reset_vector(void)
+void __bare_init __naked barebox_arm_reset_vector(void)
 {
 	arm_cpu_lowlevel_init();
 
 	beaglebone_board_init();
 
 	barebox_arm_entry(0x80000000, SZ_256M, 0);
+	/* barebox_arm_entry(0x402F0400, 0x4030CE00-0x402F0400, 0); */ /* for SRAM */
 }
-- 
1.7.10.4

