From: Jan Luebbe <jluebbe@debian.org>
Date: Fri, 29 Mar 2013 13:40:36 +0100
Subject: [PATCH] Release distrokit/beaglebone/20130329

Signed-off-by: Jan Luebbe <jluebbe@debian.org>
---
 Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/Makefile b/Makefile
index e20f162..ac5933b 100644
--- a/Makefile
+++ b/Makefile
@@ -1,7 +1,7 @@
 VERSION = 3
 PATCHLEVEL = 8
 SUBLEVEL = 4
-EXTRAVERSION =
+EXTRAVERSION =-20130329
 NAME = Unicycling Gorilla
 
 # *DOCUMENTATION*
